gazebo_ros_control/src/default_robot_hw_sim.cpp

```diff
@@ -0,0 +1,654 @@
+        if (physics_type_.compare("ode") == 0)
+        {
+          sim_joints_[j]->SetParam("vel", 0, e_stop_active_ ? 0 : joint_velocity_command_[j]);
+        }
+        else 
+        {
+          sim_joints_[j]->SetVelocity(0, e_stop_active_ ? 0 : joint_velocity_command_[j]);
+        }
+#else
+        sim_joints_[j]->SetVelocity(0, e_stop_active_ ? 0 : joint_velocity_command_[j]);
+#endif
+        break;
+
+      case VELOCITY_PID:
+        double error;
+        if (e_stop_active_)
+          error = -joint_velocity_[j];
+        else
+          error = joint_velocity_command_[j] - joint_velocity_[j];
+        const double effort_limit = joint_effort_limits_[j];
+#if 0 //@todo
+        const double effort = clamp(pid_controllers_[j].computeCommand(error, period),
+                                    -effort_limit, effort_limit);
+#else
+        const double effort = 0.0;
+#endif
+        sim_joints_[j]->SetForce(0, effort);
+        break;
+    }
+  }
+}
+
+void DefaultRobotHWSim::eStopActive(const bool active)
+{
+  e_stop_active_ = active;
+}
+
+// Register the limits of the joint specified by joint_name and joint_handle. The limits are
+// retrieved from joint_limit_nh. If urdf_model is not NULL, limits are retrieved from it also.
+// Return the joint's type, lower position limit, upper position limit, and effort limit.
+void DefaultRobotHWSim::registerJointLimits(const std::string& joint_name,
+                         const hardware_interface::JointStateHandle& joint_handle,
+                         const ControlMethod ctrl_method,
+                         const rclcpp::Node::SharedPtr& joint_limit_nh,
+                         const urdf::Model *const urdf_model,
+                         int *const joint_type, double *const lower_limit,
+                         double *const upper_limit, double *const effort_limit,
+                         double *const vel_limit)
+{
+  *joint_type = urdf::Joint::UNKNOWN;
+  *lower_limit = -std::numeric_limits<double>::max();
+  *upper_limit = std::numeric_limits<double>::max();
+  *effort_limit = std::numeric_limits<double>::max();
+
+  joint_limits_interface::JointLimits limits;
+  bool has_limits = false;
+
+  joint_limits_interface::SoftJointLimits soft_limits;
+  bool has_soft_limits = false;
+
+  if (urdf_model != NULL)
+  {
+    const urdf::JointConstSharedPtr urdf_joint = urdf_model->getJoint(joint_name);
+    if (urdf_joint != NULL)
+    {
+      *joint_type = urdf_joint->type;
+      // Get limits from the URDF file.
+      if (joint_limits_interface::getJointLimits(urdf_joint->name, joint_limit_nh, limits))
+        has_limits = true;
+      if (joint_limits_interface::getSoftJointLimits(urdf_joint->name, joint_limit_nh, soft_limits))
+        has_soft_limits = true;
+      
+      //@note (ddeng): these joint limits arent input into the node, so fetch them from the urdf
+      limits.min_position = urdf_joint->limits->lower;
+      limits.max_position = urdf_joint->limits->upper;
```

Comment:
This function DefaultRobotHWSim::registerJointLimits needs some trimming; it has a bunch of repeat logic.